dynamic/clientset/dynamic_clientset.go

```diff
@@ -72,33 +81,27 @@ func (cs *Clientset) resource(apiResource *dynamicdiscovery.APIResource, namespa
 	if gv.Group != "" {
 		config.APIPath = "/apis"
 	}
-	dc, err := dynamic.NewClient(&config)
+	dc, err := dynamic.NewForConfig(&config)
 	if err != nil {
 		return nil, fmt.Errorf("can't create dynamic client for resource %v in apiVersion %v: %v", apiResource.Name, apiResource.APIVersion, err)
 	}
+	resourceGVR := getResourceGVR(apiResource)
 	return &ResourceClient{
-		ResourceInterface: dc.Resource(&apiResource.APIResource, namespace),
-		dc:                dc,
-		gv:                gv,
-		resource:          apiResource,
+		NamespaceableResourceInterface: dc.Resource(resourceGVR),
+		dc:                             dc,
+		gv:                             gv,
+		resource:                       apiResource,
 	}, nil
 }
 
 type ResourceClient struct {
-	dynamic.ResourceInterface
+	dynamic.NamespaceableResourceInterface
 
-	dc       *dynamic.Client
+	dc       dynamic.Interface
 	gv       schema.GroupVersion
 	resource *dynamicdiscovery.APIResource
 }
 
-func (rc *ResourceClient) WithNamespace(namespace string) *ResourceClient {
```

Comment:
In order to add `UpdateWithRetries()` for namespaced clients, we probably need to override `Namespace()` here and make it return a new wrapper struct that anonymously embeds a `dynamic.ResourceInterface` as well as implementing `UpdateWithRetries()`.

Reply:
Hasn't `UpdateWithRetries()` implemented already [here](https://github.com/GoogleCloudPlatform/metacontroller/blob/master/dynamic/clientset/dynamic_clientset.go#L125-L145)?

Reply:
We need to implement UpdateWithRetries() twice, although they can both just call the same private function that does the real work. My goal was to match the pattern used in the new dynamic client for handling the fact that some API calls are namespaced and others are not.

Notice that [NamespaceableResourceInterface](https://godoc.org/k8s.io/client-go/dynamic#NamespaceableResourceInterface) contains both a `Namespace()` method *and* a full embedding of ResourceInterface:

```go
type NamespaceableResourceInterface interface {
    Namespace(string) ResourceInterface
    ResourceInterface
}
```

That means, for example, you can do both `dc.Resource().List()` and `dc.Resource().Namespace().List()`. Which one you use depends on whether you want to send a namespaced request or a cluster-scoped request. For example, you can list objects for one namespace, or for all namespaces.

To match this for the extra functions we add like UpdateWithRetries(), we should make sure both `rc.UpdateWithRetries()` and `rc.Namespace().UpdateWithRetries()` work, because we want to be able to update objects that are either cluster-scoped or namespaced.

Reply:
Oh, i see! I have to support rc.UpdateWithRetries() too for cluster-scoped operations.